pkg/uhttp/dbcache.go

```diff
@@ -0,0 +1,674 @@
+	}
+
+	cacheKey := fmt.Sprintf("%x", hash.Sum(nil))
+	return cacheKey, nil
+}
+
+func (d *DBCache) Load(ctx context.Context) (*DBCache, error) {
+	l := ctxzap.Extract(ctx)
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		l.Debug("Failed to read user cache directory", zap.Error(err))
+		return &DBCache{}, err
+	}
+
+	// Connect to db
+	sqlDB, err := sql.Open("sqlite3", filepath.Join(cacheDir, "lcache.db"))
+	if err != nil {
+		l.Debug("Failed to open database", zap.Error(err))
+		return &DBCache{}, err
+	}
+
+	d.db = sqlDB
+	return d, nil
+}
+
+// Get returns cached response (if exists).
+func (d *DBCache) Get(ctx context.Context, key string) (*http.Response, error) {
+	if d.IsNilConnection() {
+		return nil, fmt.Errorf("%s", nilConnection)
+	}
+
+	entry, err := d.Select(ctx, key)
+	if err == nil && len(entry) > 0 {
+		r := bufio.NewReader(bytes.NewReader(entry))
+		resp, err := http.ReadResponse(r, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		err = d.Hits(ctx, key)
+		if err != nil {
+			ctxzap.Extract(ctx).Debug("Failed to update hits", zap.Error(err))
+		}
+
+		return resp, nil
+	}
+
+	err = d.Misses(ctx, key)
+	if err != nil {
+		ctxzap.Extract(ctx).Debug("Failed to update misses", zap.Error(err))
+	}
+
+	return nil, nil
+}
+
+// Set stores and save response in the db.
+func (d *DBCache) Set(ctx context.Context, key string, value *http.Response) error {
+	var url string
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	cacheableResponse, err := httputil.DumpResponse(value, true)
+	if err != nil {
+		return err
+	}
+
+	if value.Request != nil {
+		url = getFullUrl(value.Request)
+	}
+
+	err = d.Insert(ctx,
+		key,
+		cacheableResponse,
+		url,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Remove stored keys.
+func (d *DBCache) Delete(ctx context.Context, key string) error {
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	err := d.Remove(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *DBCache) cleanup(ctx context.Context) error {
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	stats, err := d.getStats(ctx)
+	if err != nil {
+		l.Debug("error getting stats", zap.Error(err))
+		return err
+	}
+
+	l.Debug("summary and stats", zap.Any("stats", stats))
+	err = d.close(ctx)
+	if err != nil {
+		l.Debug("error closing db", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+// Insert data into the cache table.
+func (d *DBCache) Insert(ctx context.Context, key string, value any, url string) error {
+	var (
+		bytes []byte
+		err   error
+		ok    bool
+	)
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	if bytes, ok = value.([]byte); !ok {
+		bytes, err = json.Marshal(value)
+		if err != nil {
+			l.Debug("error marshaling data", zap.Error(err))
+			return err
+		}
+	}
+
+	if ok, _ := d.Has(ctx, key); !ok {
```

Comment:
You need to do the select to check for an existing key after starting the transaction, not before. Otherwise you have a race condition.

Reply:
Not quite sure about it, but let me check it